lib/runBlock.js

```diff
@@ -32,237 +32,187 @@ module.exports = function (opts, cb) {
-   * @private
-   * @param {Function} cb the callback is given error if there are any
+   * The `afterBlock` event
+   *
+   * @event Event: afterBlock
+   * @type {Object}
+   * @property {Object} result emits the results of processing a block
    */
-  function processTransactions (cb) {
-    var validReceiptCount = 0
-
-    async.eachSeries(block.transactions, processTx, cb)
-
-    function processTx (tx, cb) {
-      var gasLimitIsHigherThanBlock = new BN(block.header.gasLimit).lt(new BN(tx.gasLimit).add(gasUsed))
-      if (gasLimitIsHigherThanBlock) {
-        cb(new Error('tx has a higher gas limit than the block'))
-        return
-      }
-
-      // run the tx through the VM
-      self.runTx({
-        tx: tx,
-        block: block
-      }, parseTxResult)
-
-      function parseTxResult (err, result) {
-        txResults.push(result)
-        // var receiptResult = new BN(1)
-
-        // abort if error
-        if (err) {
-          receipts.push(null)
-          cb(err)
-          return
-        }
+  await this._emit('afterBlock', { receipts: result.receipts, results: result.results })
 
-        gasUsed = gasUsed.add(result.gasUsed)
-        // combine blooms via bitwise OR
-        bloom.or(result.bloom)
-
-        if (generateStateRoot) {
-          block.header.bloom = bloom.bitvector
-        }
-
-        var txLogs = result.vm.logs || []
-
-        var rawTxReceipt = [
-          result.vm.exception ? 1 : 0, // result.vm.exception is 0 when an exception occurs, and 1 when it doesn't.  TODO make this the opposite
-          gasUsed.toArrayLike(Buffer),
-          result.bloom.bitvector,
-          txLogs
-        ]
-        var txReceipt = {
-          status: rawTxReceipt[0],
-          gasUsed: rawTxReceipt[1],
-          bitvector: rawTxReceipt[2],
-          logs: rawTxReceipt[3]
-        }
+  return result
+}
 
-        receipts.push(txReceipt)
-        receiptTrie.put(rlp.encode(validReceiptCount), rlp.encode(rawTxReceipt), function () {
-          validReceiptCount++
-          cb()
-        })
-      }
+/**
+ * Validates and applies a block, computing the results of
+ * applying its transactions. This method doesn't modify the
+ * block itself. It computes the block rewards and puts
+ * them on state (but doesn't persist the changes).
+ * @param {Block} block
+ * @param {Boolean} [skipBlockValidation=false]
+ */
+async function applyBlock (block, skipBlockValidation = false) {
+  const state = new PStateManager(this.stateManager)
+  // Validate block
+  if (!skipBlockValidation) {
+    if (new BN(block.header.gasLimit).gte(new BN('8000000000000000', 16))) {
+      throw new Error('Invalid block with gas limit greater than (2^63 - 1)')
+    } else {
+      await promisify(block.validate).bind(block)(this.blockchain)
     }
   }
+  // Apply transactions
+  const txResults = await applyTransactions.bind(this)(block, state)
+  // Pay ommers and miners
+  await assignBlockRewards.bind(this)(block)
+  return txResults
+}
```

Comment:
Moved ``validateBlock``, ``processTransactions`` and ``payOmmersAndMiners`` here, introduced hierarchical call structure for ``applyBlock`` and ``applyTransactions``, good.